### PR TITLE
Clean up evaporate_sublimate_precip

### PR DESCRIPTION
**Pull request by PeterCaldwell**
Fixes a bug in precip evaporation which allowed clear-sky qv to become negative, resulting in overly strong evaporation. Also generally cleans up evaporation calculation and ensures it stays within reasonable bounds.

### CODE REVIEW

**Comment by PeterCaldwell** [comment]
I think this PR is ready for integration even though I'm not totally happy with it. Future work:

- p3_main's unit test has negative qsat and qv and T=800K, but I think that's the unit test's fault. 2 day ne30 runs produce reasonable values
- qv can be supersaturated after evap_precip even though inclusion of "ab" is supposed to approximately fix this
- should saturation adjustment use ab computed by get_time_space_phys_variables rather than calculating psychrometric correction from scratch?
- I'd really rather not have a cell-ave saturation adjustment call at the beginning of P3 and I'm not sure we don't need one at the end of P3.

Basically I'm unhappy with the entire evap_precip conceptual framework. I'll try to fix these things in future work, but want something non-erroneous for the time being.

File: components/cam/src/physics/cam/micro_p3.F90 [hunk 1]
```diff
@@ -3188,39 +3215,68 @@ subroutine evaporate_sublimate_precip(qr_incld,qc_incld,nr_incld,qitot_incld,
...
 
-      ! rain evaporation
-      if (qr_incld.ge.qsmall) then
-         qrevp = epsr * (qclr-qvs)/ab
-      end if
+      ! compute in-evaporating-region evap rate
+      ! Note qclr<=qv<=qvs so qclr-qvs should be negative.
+      ! In practice roundoff error *might* make qrevp slightly negative, so adding max(0,...)
+      ! The minus sign in front makes qrevp positive
+      qrevp = max(0._rtype, -epsr * (qclr-qvs)/ab )
```
Comment (JS-WRF-SBM) [inline components/cam/src/physics/cam/micro_p3.F90:3264]: @PeterCaldwell 

- The rain `qrevp` depend here on 'pushing' the clear-sky area to the level of saturation (which is a large value; in reality it will be hard getting a volume of air saturated just by evaporation).

- The `epsr` factor should of course bound it. However, looking at the WRF version, this term is multiplying by a term `oxx*(1-exp(-xx*dt))`. `oxx` is the total relaxation time (cloud, rain, ice). The total relaxation time involves info about `qc,nc,qr,nr,qi,ni` and mean sizes - thus getting the CLUBB changes directly (explicit). 
My point here is that according to the current formulation, you might get a large number which still needs to be bounded (factored).

- Possibly we can discuss it offline, but referring to the `qrevp` WRF version: assuming that the best we can do is to use the temperature and water vapor tendencies (e.g., T - T_old, qv - qv_old) which stores the current time step changes in condensation/evaporation happen in CLUBB - is there additional in-consistency you this will misrepresent `qrevp`?

File: components/cam/src/physics/cam/micro_p3.F90 [hunk 2]
```diff
@@ -3188,39 +3215,68 @@ subroutine evaporate_sublimate_precip(qr_incld,qc_incld,nr_incld,qitot_incld,
...
+   !    print*
+   !    call endscreamrun("qv supersaturated in evap_precip")
+   ! endif
+   
+   ! Determine temporary cloud fraction which ramps to 0 as cloud water
+   ! approaches zero. This avoids problems due to lcldm having a positive
+   ! minimum value which would result in evaporation never occurring in a
+   ! small portion of the grid. Ramping rather than thresholding is used
+   ! to preserve convergence.
+   if (qc_incld < 1.e-6_rtype) then
```
Comment (JS-WRF-SBM) [inline components/cam/src/physics/cam/micro_p3.F90:3234]: @PeterCaldwell 

Would you explain how convergence affects here?
You chose relatively high cloud droplet mass to represent a cloudy area.

Comment (PeterCaldwell) [reply components/cam/src/physics/cam/micro_p3.F90:3234]: Suppose qc_incld is infinitesimally bigger than 1e-6 at coarser resolution and drops to infinitesimally smaller than 1e-6 at higher resolution. All of a sudden the code would be handled quite differently if no ramp were present. This often causes jumps on convergence plots. I'd like to always use ramps at all thresholds but some cases are hard to handle. Does that make sense? Should I change the comment to be more clear? And yes, 1e-6 is a fairly large threshold. The fact that we're ramping instead of thresholding below that probably makes it less important, but a smaller number is worth testing.

File: components/cam/src/physics/cam/micro_p3.F90 (hunk 1 above)
Comment (PeterCaldwell) [reply components/cam/src/physics/cam/micro_p3.F90:3264]: Interesting point. It is hard to imagine that any parameterization of rain evaporation won't boil down to the difference between clear-sky and saturated air scaled by the supersaturation removal timescale epsr and accounting for the fact that evaporation cooling increases qsat, damping evaporative strength (ab in the parameterization above). I took epsr from the p3 table lookup without thinking about it, but perhaps this version of epsr is inconsistent with our current implementation of qvclr - qsat?

Comment (JS-WRF-SBM) [reply components/cam/src/physics/cam/micro_p3.F90:3264]: I think the formulation for `epsr` is similar to the one done in WRF-p3.
Your `qrevp` formulation seems to lack additional scale with the actual MP time-step (`dt` or `odt`) and total relation time (`xx` or 'oxx') as per the original WRF-p3 scheme.

The formulation would have looked then like:   
`qrevp = (epsr * (qclr-qvs) * odt*oxx*(1.0-exp(-xx*dt))) / ab` 

And again, this is only a single term (out of 3) in the WRF-p3, so not sure the magnitude would be bounded ...

File: components/cam/src/physics/cam/micro_p3.F90 (hunk 2 above)
Comment (JS-WRF-SBM) [reply components/cam/src/physics/cam/micro_p3.F90:3234]: Thanks for the explanation.

**Comment by AaronDonahue** [comment]
@PeterCaldwell , what's the status on this PR?  It's been open for over a month.

**Comment by PeterCaldwell** [comment]
"what's the status on this PR?": @crterai will issue a new PR soon which builds on this one and fixes some problems with it. We will abandon this PR.
